tracer/src/Datadog.Trace/RemoteConfigurationManagement/RemoteConfigurationManager.cs

```diff
@@ -0,0 +1,290 @@
+            IDiscoveryService discoveryService,
+            IRemoteConfigurationApi remoteConfigurationApi,
+            RemoteConfigurationSettings settings,
+            string serviceName)
+        {
+            SemaphoreSlim.Wait();
+
+            try
+            {
+                Instance ??= new RemoteConfigurationManager(
+                    discoveryService,
+                    remoteConfigurationApi,
+                    id: settings.Id,
+                    runtimeId: settings.RuntimeId,
+                    tracerVersion: settings.TracerVersion,
+                    service: serviceName,
+                    env: settings.Environment,
+                    appVersion: settings.AppVersion,
+                    pollInterval: settings.PollInterval);
+            }
+            finally
+            {
+                SemaphoreSlim.Release();
+            }
+
+            return Instance;
+        }
+
+        public async Task StartPollingAsync()
+        {
+            await SemaphoreSlim.WaitAsync().ConfigureAwait(false);
+
+            if (_isPollingStarted)
+            {
+                Log.Warning("Remote Configuration management polling is already started.");
+                return;
+            }
+
+            _isPollingStarted = true;
+            LifetimeManager.Instance.AddShutdownTask(OnShutdown);
+
+            SemaphoreSlim.Release();
+
+            while (!_cancellationSource.IsCancellationRequested)
+            {
+                var isRcmEnabled = !string.IsNullOrEmpty(_discoveryService.ConfigurationEndpoint);
+                var isProductRegistered = _products.Any();
+
+                if (isRcmEnabled && isProductRegistered)
+                {
+                    await Poll().ConfigureAwait(false);
+                    _lastPollError = null;
+                }
+
+                try
+                {
+                    await Task.Delay(_pollInterval, _cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (TaskCanceledException)
+                {
+                    // We are shutting down, so don't do anything about it
+                }
+            }
+        }
+
+        public void RegisterProduct(Product product)
+        {
+            _products.TryAdd(product.Name, product);
+        }
+
+        public void UnregisterProduct(string productName)
+        {
+            _products.TryRemove(productName, out _);
+        }
+
+        private async Task Poll()
+        {
+            await SemaphoreSlim.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                var request = BuildRequest();
+                var response = _remoteConfigurationApi.GetConfigs(request).Result;
+
+                if (response?.Targets?.Signed != null)
+                {
+                    ProcessResponse(response);
+                    _targetsVersion = response.Targets.Signed.Version;
+                }
```

Comment:
I"m not sure in what case it could happen but should we add a debug log in an else clause here?

Reply:
Well, it happens time to time.
If nothing changes, our backend can return empty body.